File: src/RepositoryModules1/RepositoryModule1.vb
﻿Imports System.IO
Imports System.Text
Imports System.Threading
Imports Storex

<RepositoryModuleExport("E0B3F83A-B417-43DB-8CCF-9916A2EB63C6", "簡易ファイルシステム", Description:="モードでシンボル内容等を設定します")>
Public Class RepositoryModule1
    Implements IRepositoryModule

    Friend ReadOnly TextEncoding As Encoding = Encoding.GetEncoding("shift_jis")

    Dim CurrentMode As IMode
    Dim CurrentUser As IUser

    Public ReadOnly Property IsConfiguable As Boolean = True Implements IRepositoryModule.IsConfiguable

    Public ReadOnly Property IsModeConfiguable As Boolean = True Implements IRepositoryModule.CanConfigureMode

    Public Function ConfigureAsync() As Task Implements IModule.ConfigureAsync

        ConfigForm1.Configure()

        Return Task.CompletedTask
    End Function

    Public Function ConfigureModeAsync(Mode As IMode) As Task Implements IRepositoryModule.ConfigureModeAsync

        ModeConfigForm1.Configure(Mode)

        Return Task.CompletedTask
    End Function

    Public Function PrepareAsync(Mode As IMode, UserInfo As IUser) As Task Implements IRepositoryModule.PrepareAsync

        If Not File.Exists(My.Settings.FilePath) Then

            Throw New RepositoryModuleException("概要データ ファイルが正しく設定されていません。")
        End If

        If Not Directory.Exists(My.Settings.FolderPath) Then

            Throw New RepositoryModuleException("詳細データ フォルダが正しく設定されていません。")
        End If

        If Not LabelDefinition1.TryExtract(Mode, LabelDefinition1.KeyForPrimary) Then

            Throw New RepositoryModuleException("モードが構成されていません。")
        End If

        CurrentMode = Mode
        CurrentUser = UserInfo

        Return Task.CompletedTask
    End Function

    Public Function FindPrimaryLabel(Sources() As ILabelSource) As ILabel Implements IRepositoryModule.FindPrimaryLabel

        Dim MatchSymbols = Sources _
            .OfType(Of Symbol) _
            .Select(
                Function(x)

                    Dim LabelDefinition As LabelDefinition1 = Nothing
                    Dim Label As BasicLabel = Nothing
                    Dim IsMatch = LabelDefinition1.TryExtract(CurrentMode, LabelDefinition1.KeyForPrimary, LabelDefinition) AndAlso LabelDefinition.TryGeneraLabel(x, Label)

                    Return New With {IsMatch, Label}
                End Function
                ) _
            .Where(Function(x) x.IsMatch) _
            .Select(Function(x) x.Label) _
            .ToArray()

        Return If(MatchSymbols.Length = 1, MatchSymbols(0), Nothing)
    End Function

    Public Function FindSecondaryLabels(Primary As ILabel, Sources() As ILabelSource) As ILabel() Implements IRepositoryModule.FindSecondaryLabels

        If CurrentMode.HasSecondaryLabel Then

            Dim MatchSymbols = Sources _
                .OfType(Of C3Label) _
                .Where(Function(x) x.PartNumber = Primary.ItemNumber) _
                .ToArray()

            Return MatchSymbols
        End If

        Return Array.Empty(Of ILabel)
    End Function

    Public Function RegisterAsync(Primary As ILabel, Secondary As ILabel, CaptureDatas() As CaptureData, Tags() As String) As Task Implements IRepositoryModule.RegisterAsync

        Return RegisterAsync(Primary, Secondary, CaptureDatas, Tags, CancellationToken.None)
    End Function

    Public Async Function RegisterAsync(Primary As ILabel, Secondary As ILabel, CaptureDatas() As CaptureData, Tags() As String, cancellationToken As CancellationToken) As Task Implements IRepositoryModule.RegisterAsync

        Dim Timestamp As Date = Now

        Dim MyFile = New FileInfo(My.Settings.FilePath)

        Using SW = New StreamWriter(MyFile.FullName, True, TextEncoding)

            If MyFile.Length = 0 Then

                Await SW.WriteLineAsync(String.Join(vbTab, New String() {
                    "PrimaryPartNumber",
                    "PrimarySerialNumber",
                    "SecondaryPartNumber",
                    "SecondarySerialNumber",
                    "ModeId",
                    "UserName",
                    "Timestamp"
                    }))
            End If

            Await SW.WriteLineAsync(String.Join(vbTab, New String() {
                Primary.ItemNumber,
                Primary.SerialNumber,
                Secondary?.ItemNumber,
                Secondary?.SerialNumber,
                CurrentMode.Id.ToString(),
                CurrentUser?.DisplayName,
                $"{Timestamp:yyyy/MM/dd HH:mm:ss.fff}"
                }))
        End Using

        Dim MyFolder = New DirectoryInfo(My.Settings.FolderPath)

        MyFolder = MyFolder.CreateSubdirectory($"{Timestamp:yyyy-MM-dd HH-mm-ss.fff}")

        For i = 1 To CaptureDatas.Count

            Dim CaptureData As CaptureData = CaptureDatas(i - 1)

            Dim CaptureDataFolder = MyFolder.CreateSubdirectory($"catpure#{i}")

            File.WriteAllBytes($"{CaptureDataFolder.FullName}\original.jpg", CaptureData.OriginalImage)

            If CaptureData.AdornedImage IsNot Nothing Then

                File.WriteAllBytes($"{CaptureDataFolder.FullName}\adorned.jpeg", CaptureData.AdornedImage)
            End If

            If CaptureData.LabelSources.Count > 0 Then

                Dim SymbolValues() As String = CaptureData.LabelSources.SelectMany(Function(x) x.Symbols.Select(Function(xx) xx.Value)).ToArray()

                File.WriteAllLines($"{CaptureDataFolder.FullName}\symbols.txt", SymbolValues)
            End If
        Next

        If Tags.Length > 0 Then

            File.WriteAllLines($"{MyFolder.FullName}\tags.txt", Tags)
        End If
    End Function

    Public Sub Dispose() Implements IDisposable.Dispose
    End Sub
End Class

File: src/RepositoryModules.CS/RepositoryModule1.cs
﻿using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Storex;

namespace RepositoryModules
{
    using static RepositoryModuleHelper;

    [RepositoryModuleExport("E0B3F83A-B417-43DB-8CCF-9916A2EB63C6", "簡易ファイルシステム", Description = "モードでシンボル内容等を設定します")]
    public class RepositoryModule1 : IRepositoryModule
    {
        static readonly Encoding _textEncoding = Encoding.GetEncoding("shift_jis");

        IMode _currentMode;
        IUser _currentUser;

        public bool IsConfiguable { get; private set; } = true;

        public bool IsModeConfiguable { get; private set; } = true;

        public bool CanConfigureMode => true;

        public Task ConfigureAsync(CancellationToken cancellationToken)
        {
            ConfigForm1.Configure();

            return Task.CompletedTask;
        }

        public Task ConfigureModeAsync(IMode mode, CancellationToken cancellationToken)
        {
            ModeConfigForm1.Configure(mode);

            return Task.CompletedTask;
        }

        public Task<bool> PrepareAsync(IMode Mode, IUser UserInfo, CancellationToken cancellationToken)
        {
            if (!File.Exists(MySettings.Default.FilePath))
            {
                throw new RepositoryModuleException("概要データ ファイルが正しく設定されていません。");
            }

            if (!Directory.Exists(MySettings.Default.FolderPath))
            {
                throw new RepositoryModuleException("詳細データ フォルダが正しく設定されていません。");
            }

            if (!Mode.TryExtractProperty(FixedLengthSpec.PropertyKeyForPrimary, out _priaryLabelFixedLengthSpec) |
                !Mode.TryExtractProperty(SecondaryLabelCriteria.PropertyKey, out _secondaryLabelCriteria))
            {
                throw new RepositoryModuleException("モードが構成されていません。");
            }

            _conditionDictionary = ReadSecondaryConditionsFile();

            _currentMode = Mode;
            _currentUser = UserInfo;

            return Task.FromResult(true);
        }

        FixedLengthSpec _priaryLabelFixedLengthSpec;

        public Task<ILabel> FindPrimaryLabelAsync(ILabelSource[] sources, CancellationToken cancellationToken)
        {
            var labels = sources
                .OfType<Symbol>()
                .Select(x => _priaryLabelFixedLengthSpec.TryGenerateLabel(x, out var label) ? label : null)
                .Where(x => x != null)
                .ToArray();

            return Task.FromResult(labels.Length == 1 ? labels[0] : null);
        }

        SecondaryLabelCriteria _secondaryLabelCriteria;
        IReadOnlyDictionary<string, IEnumerable<string>> _conditionDictionary;

        public Task<ILabel[]> FindSecondaryLabelsAsync(ILabel primary, ILabelSource[] sources, CancellationToken cancellationToken)
        {
            var labels = new List<ILabel>();

            var validConditionValues = 
                _secondaryLabelCriteria.SpecifiedByConditionFileBehavior != SecondaryLabelBehavior.Deny
                    &&  _conditionDictionary.TryGetValue(primary.ItemNumber, out var strings) ? 
                strings :
                Array.Empty<string>();

            var c3Labels = sources
                .Where(_ => (_secondaryLabelCriteria.AcceptableTypes & SecondaryLabelTypes.C3Label) == SecondaryLabelTypes.C3Label)
                .OfType<C3Label>()
                .Where(o =>
                    (_secondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior != SecondaryLabelBehavior.Deny && EqualsIgnoreCase(o.PartNumber, primary.ItemNumber))
                    | (_secondaryLabelCriteria.SpecifiedByConditionFileBehavior != SecondaryLabelBehavior.Deny && validConditionValues.Any(p => EqualsIgnoreCase(o.PartNumber, p)))
                    | (_secondaryLabelCriteria.OtherNotSingleSymbolLabelsBehavior != SecondaryLabelBehavior.Deny && !EqualsIgnoreCase(o.PartNumber, primary.ItemNumber) & !validConditionValues.Any(p => EqualsIgnoreCase(o.PartNumber, p)))
                    )
                .ToList();

            labels.AddRange(c3Labels);

            var singleSymbolLabels = sources
                .Where(_ => (_secondaryLabelCriteria.AcceptableTypes & SecondaryLabelTypes.SingleSymbol) == SecondaryLabelTypes.SingleSymbol)
                .OfType<Symbol>()
                .Where(x => !_priaryLabelFixedLengthSpec.TryGenerateLabel(x, out _))
                .Select(o => new BasicLabel(o) { ItemNumber = o.Value })
                .Where(o =>
                    (_secondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior != SecondaryLabelBehavior.Deny & EqualsIgnoreCase(o.ItemNumber, primary.ItemNumber))
                    | (_secondaryLabelCriteria.SpecifiedByConditionFileBehavior != SecondaryLabelBehavior.Deny & validConditionValues.Any(p => EqualsIgnoreCase(o.ItemNumber, p)))
                    )
                .ToList();

            labels.AddRange(singleSymbolLabels);

            return Task.FromResult(labels.ToArray());
        }

        public async Task<bool> RegisterAsync(ILabel primary, ILabel secondary, CaptureData[] captureDatas, string[] tags, CancellationToken cancellationToken)
        {
            if (!File.Exists(MySettings.Default.FilePath))
            {
                throw new RepositoryModuleException("概要データ ファイルが存在しません。");
            }

            if (!Directory.Exists(MySettings.Default.FolderPath))
            {
                throw new RepositoryModuleException("詳細データ フォルダが存在しません。");
            }

            if (secondary is null)
            {
                if (_secondaryLabelCriteria.NoLabelBehavior == SecondaryNoLabelBehavior.Warnning
                    && !ShowAlert("セカンダリ ラベルが必要です。無視して登録しますか。"))
                {
                    return false;
                }
                else if (_secondaryLabelCriteria.NoLabelBehavior == SecondaryNoLabelBehavior.Deny)
                {
                    ShowError("セカンダリ ラベルが必要です。");
                    return false;
                }
                else if (_secondaryLabelCriteria.NoLabelBehavior != SecondaryNoLabelBehavior.Default)
                {
                    var hasItemNumberEqualsToPrimary = tags.Any(o => EqualsIgnoreCase(o, primary.ItemNumber));
                    var hasSpecifiedByConditionFile = 
                        tags.Any(o => _conditionDictionary.TryGetValue(primary.ItemNumber, out var strings) && strings.Any(p => EqualsIgnoreCase(o, p)));

                    if (_secondaryLabelCriteria.NoLabelBehavior == SecondaryNoLabelBehavior.WarningWhenTagNotMatched
                        && !hasItemNumberEqualsToPrimary && !hasSpecifiedByConditionFile
                        && !ShowAlert("セカンダリ ラベル、又は対応するタグが必要です。無視して登録しますか。"))
                    {
                        return false;
                    }
                    else if (_secondaryLabelCriteria.NoLabelBehavior == SecondaryNoLabelBehavior.DenyWhenTagNotMatched
                        && !hasItemNumberEqualsToPrimary && !hasSpecifiedByConditionFile)
                    {
                        ShowError("セカンダリ ラベル、又は対応するタグが必要です。");
                        return false;
                    }
                    else if (_secondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior == SecondaryLabelBehavior.Warnning
                        && hasItemNumberEqualsToPrimary
                        && !ShowAlert("プライマリラベルと同じ品番がタグに含まれています。登録しますか。"))
                    {
                        return false;
                    }
                    else if (_secondaryLabelCriteria.SpecifiedByConditionFileBehavior == SecondaryLabelBehavior.Warnning
                        && hasSpecifiedByConditionFile
                        && !ShowAlert("対応表で指定されている品番がタグに含まれています。登録しますか。"))
                    {
                        return false;
                    }
                }
            }
            else if (_secondaryLabelCriteria.ItemNumberEqualsToPrimaryOneBehavior == SecondaryLabelBehavior.Warnning
                && EqualsIgnoreCase(primary.ItemNumber, secondary.ItemNumber)
                && !ShowAlert("指定された セカンダリ ラベルは、プライマリラベルと同じ品番です。登録を続行しますか。"))
            {
                return false;
            }
            else if (_secondaryLabelCriteria.SpecifiedByConditionFileBehavior == SecondaryLabelBehavior.Warnning
                && _conditionDictionary.TryGetValue(primary.ItemNumber, out var strings1) 
                && strings1.Any(o => EqualsIgnoreCase(o, secondary.ItemNumber))
                && !ShowAlert("指定された セカンダリ ラベルは、プライマリ ラベルとは異なり対応表と同じ品番です。登録を続行しますか。"))
            {
                return false;
            }
            else if (_secondaryLabelCriteria.OtherNotSingleSymbolLabelsBehavior == SecondaryLabelBehavior.Warnning
                && !EqualsIgnoreCase(primary.ItemNumber, secondary.ItemNumber)
                && !(_conditionDictionary.TryGetValue(primary.ItemNumber, out var strings2) && strings2.Any(o => EqualsIgnoreCase(o, secondary.ItemNumber)))
                && !ShowAlert("指定された セカンダリ ラベルは、プライマリラベルや対応表と異なる品番です。登録しますか。"))
            {
                return false;
            }

            var timestamp = DateTime.Now;

            var myFile = new FileInfo(MySettings.Default.FilePath);

            using (var sw = new StreamWriter(myFile.FullName, true, _textEncoding))
            {
                if (myFile.Length == 0L)
                {
                    await sw.WriteLineAsync(string.Join("\t", new string[]
                    {
                        "PrimaryPartNumber",
                        "PrimarySerialNumber",
                        "SecondaryPartNumber",
                        "SecondarySerialNumber",
                        "ModeId",
                        "UserName",
                        "Timestamp",
                    }));
                }

                await sw.WriteLineAsync(string.Join("\t", new string[]
                {
                    primary.ItemNumber,
                    primary.SerialNumber,
                    secondary?.ItemNumber,
                    secondary?.SerialNumber,
                    _currentMode.Id.ToString(),
                    _currentUser?.DisplayName,
                    $"{timestamp:yyyy/MM/dd HH:mm:ss.fff}",
                }));
            }

            var mFolder = new DirectoryInfo(MySettings.Default.FolderPath);

            mFolder = mFolder.CreateSubdirectory($"{timestamp:yyyy-MM-dd HH-mm-ss.fff}");

            for (int i = 1, loopTo = captureDatas.Count(); i <= loopTo; i++)
            {
                var captureData = captureDatas[i - 1];

                var captureDataFolder = mFolder.CreateSubdirectory($"catpure#{i}");

                WriteImage($@"{captureDataFolder.FullName}\original.jpg", captureData.OriginalImage.ToArray());

                if (captureData.AdornedImage != null)
                {
                    WriteImage($@"{captureDataFolder.FullName}\adorned.jpeg", captureData.AdornedImage.ToArray());
                }

                if (captureData.LabelSources.Count > 0)
                {
                    var symbolValues = captureData.LabelSources.SelectMany(x => x.Symbols.Select(xx => xx.Value)).ToArray();

                    File.WriteAllLines($@"{captureDataFolder.FullName}\symbols.txt", symbolValues);
                }
            }

            if (tags.Length > 0)
            {
                File.WriteAllLines($@"{mFolder.FullName}\tags.txt", tags);
            }

            return true;
        }

        public void Dispose() { }
    }
}